content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+
+Sized types (like structs, enums, etc) are turned into a simple pointer. As a result, each parameter or return value that is a sized type will come out as a `i32`. The exception are Arrays and Tuples, which are both sized types, but are converted depending on whether a type uses more than 32 bits. That means that the data of types like `(u8, u8)` or `[u16; 2]` will be bitpacked into a single `i32` and passed as an immediate value. Bigger types like `(u32, u32)` or `[u8; 10]` will be passed as a pointer in the form of an `i32`, pointing at the first element. Things get even more confusing if we look at function return values: If you return an array type bigger than 32 bits, the function will get _no_ return value and instead will get an additional function parameter of type `i32`, which the function will use a pointer to a place to store the result. If a function returns a tuple, it will always turn into a function parameter, regardless of the tuple's size.
+
+Unsized types (`?Sized`), like `str`, `[u8]` or `dyn MyTrait`, are split into two values: One value is the pointer to the data, the other value is a pointer to a bit of metadata. In the case of a `str` or a slice, the metadata is the length of the data. In the case of a trait object, it’s the virtual table (or vtable), which is a list of function pointers to the individual trait function implementations. If you want to know more details about what a VTable in Rust looks like, I can recommend [this article][vtable] by Thomas Bächler. Function parameters that are unsized types are turned into _two_ parameters, one for each pointer. 
+
+If a function has an unsized type as a return value, the function gets an addition function parameter in the first position. It's the callers responsibility to pass in a pointer where the function can store a “fat pointer”. A fat pointer is a tuple of two pointers, combining the the pointers for data and metadata into one data structure.
+
+> **Octowords:** While not often used, Rust does have `u128` and `i128` as types. Just like fat pointers, these 128-bit wide types are split into two separate `i64` values.
+
+As I said, this is not something you need to know to use Rust for WebAssembly. But if you want to know more, all of this _and more_ is defined in the [C ABI] for WebAssembly that LLVM utilizes.
+
+## Module size
+
+When deploying WebAssembly on the web, the size of the WebAssembly binary matters. Every byte needs to go over the network and through the browser’s WebAssembly compiler, so a smaller binary size means less time spent waiting for the user until the WebAssembly starts working. If we build our default project from above as a release build, we get a whopping 1.7MB of WebAssembly. That does not seem right for adding two numbers.
```

Comment:
"and through the browser’s WebAssembly compiler" pedantic note: except for the data section, which is cheap copy from Wasm to memory and doesn't need to be processed. Not sure how to incorporate that into the wording here, but feel like it's worth calling out, especially because data section often can have a lot of zeroes in the middle, which contribute to the size but obviously don't matter at all neither for data transfer (which should always be compressed) nor for compilation (which will just copy them in-memory).

Maybe just say it's best to compare compressed sizes rather than raw?

Reply:
Hm, I'm torn. Especially because compilation scales with uncompressed instruction size, I am not sure i want to push towards _always_ using compressed sizes. But I see your point. I'll try and add a note

Reply:
> Hm, I'm torn. Especially because compilation scales with uncompressed instruction size

Right, if we go there, ideally you'd look just at the size of the code section (e.g. via `wasm-objdump` I guess) and compare raw sizes for those. But other than compilation speed, compressed size seems like a good general heuristic.

Reply:
Ah, and don't forget that some engines, including V8, are playing with lazy compilation of functions, which would muddle the compilation speed estimations even further.